### PR TITLE
Schedules which change pattern during Daylight Savings Time cause incorrect results at hour 23

### LINKED ISSUE #5155: DST causing unexpected result on date when schedule changes
Helpdesk ticket 10742
This is the schedule in question.

```
! Schedule: Arrêt charge E 706
Schedule:Compact, 
Arrêt charge E 706, 
Fraction,
Through: 2 Aug,
For: Weekdays Weekends Holidays WinterDesignDay SummerDesignDay,
Until: 24:00, 1,
For: AllOtherDays,
Until: 24:00, 1,
Through: 3 Aug,
For: Weekdays Weekends Holidays WinterDesignDay SummerDesignDay,
Until: 15:00, 1,
Until: 24:00, 0,
For: AllOtherDays,
Until: 24:00, 0,
```

On Aug 3, from 23:00-24:00, it should be zero, but it's showing 1.

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
Fixes #5155. When a schedule is used when DST is active, the program will look at hour 1 of the schedule when time of day = 24. This change looks to the next day's schedule for hour 1.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by rraustad** [comment]
Probable conflict with #7352 Schedule Manager Refactor.

**Comment by mjwitte** [comment]
@rraustad Actually, the issue may be resolved in #7352. I'd suggest testing the defect file with that branch to see if the problem is still there.

**Comment by rraustad** [comment]
Thought I wasn't going to get paid for today's effort...

image

File: src/EnergyPlus/ScheduleManager.cc [hunk 1]
```diff
@@ -2848,6 +2848,12 @@ namespace ScheduleManager {
             if (WhichHour <= 24) {
                 Schedule(ScheduleIndex).CurrentValue = DaySchedule(DaySchedulePointer).TSValue(TimeStep, WhichHour);
             } else if (TimeStep <= NumOfTimeStepInHour) {
+                WeekSchedulePointer = Schedule(ScheduleIndex).WeekSchedulePointer(DayOfYear_Schedule + 1);
+                if (DayOfWeek <= 7 && HolidayIndex > 0) {
+                    DaySchedulePointer = WeekSchedule(WeekSchedulePointer).DaySchedulePointer(7 + HolidayIndex);
+                } else {
+                    DaySchedulePointer = WeekSchedule(WeekSchedulePointer).DaySchedulePointer(DayOfWeek);
+                }
```
Comment (rraustad) [inline src/EnergyPlus/ScheduleManager.cc:2864]: I suggest that lines 2836-2845 be moved into each of the IF blocks (I've done that for this ELSE IF to test if it works correctly). The IF blocks are executed every time through, so there is no computational savings by having these lines outside this IF block, especially if something different needs to be done based on the conditionals. And adding extra logic on line 2838 will also cost computational time (i.e., to get the right week schedule pointer). It adds a few lines of duplicate code, but would save a little on execution time.

**Comment by rraustad** [comment]
These ASHRAE901 example files do not use the weather file daylight savings period (which appear to be non-existent in epw files).

```
  RunPeriod,
    Run Period 1,            !- Name
    No,                      !- Use Weather File Daylight Saving Period
```

and instead use:

```
  RunPeriodControl:DaylightSavingTime,
    2nd Sunday in March,     !- Start Date
    1st Sunday in November;  !- End Date
```

The schedule in question is:

```
  Schedule:Compact,
    CLOTHING_SCH,            !- Name
    Any Number,              !- Schedule Type Limits Name
    Through: 04/30,          !- Field 1
    For: AllDays,            !- Field 2
    Until: 24:00,1.0,        !- Field 3
    Through: 09/30,          !- Field 5
    For: AllDays,            !- Field 6
    Until: 24:00,0.5,        !- Field 7
    Through: 12/31,          !- Field 9
    For: AllDays,            !- Field 10
    Until: 24:00,1.0;        !- Field 11
```

I reported all schedule values in develop and this branch. Then differenced them (this branch - develop) and looked for non-zero values. I found only 1 schedule with diff's. Top 4 plots are from differences and how I narrowed this down. The bottom 2 plots compare this branch with develop.

On 4/30 and 9/30 is when the schedule values change to a new pattern. This branch picks up that the pattern has changed (a different DaySchedule) and correctly adjusts DST (see bottom 2 figures).

image

File: src/EnergyPlus/ScheduleManager.cc (hunk 1 above)
Comment (JasonGlazer) [reply src/EnergyPlus/ScheduleManager.cc:2864]: This approach makes sense to me.

File: tst/EnergyPlus/unit/ScheduleManager.unit.cc [hunk 2]
```diff
@@ -83,6 +83,97 @@ TEST_F(EnergyPlusFixture, ScheduleManager_isMinuteMultipleOfTimestep)
...
+    DataGlobals::HourOfDay = 2;
+    DataEnvironment::DSTIndicator = 0;
+    DataEnvironment::DayOfYear_Schedule = 251;
+    UpdateScheduleValues();
+    // expect 3.0 for remainder of year regardless of DST
+    EXPECT_EQ(Schedule(1).CurrentValue, 3.0);
+    DataGlobals::HourOfDay = 24;
+    DataEnvironment::DSTIndicator = 1;
+    UpdateScheduleValues();
+    EXPECT_EQ(Schedule(1).CurrentValue, 3.0);
```
Comment (JasonGlazer) [inline tst/EnergyPlus/unit/ScheduleManager.unit.cc:174]: Good unit test.

**Review by JasonGlazer — COMMENTED** [review]
I reviewed and the code change looks good.

**Comment by JasonGlazer** [comment]
@mjwitte I don't know if you have time to look at this before you leave. Let me know if you plan to.

**Comment by JasonGlazer** [comment]
@Myoldmopar did you want to review this before I merge it?

**Comment by Myoldmopar** [comment]
Yeah let me take a quick pass over it.

**Comment by Myoldmopar** [comment]
If this is working then go ahead and merge it in.  I don't have any issue with the code changes.

**Pull request merged**